stdlib/public/runtime/Casting.cpp

```diff
@@ -2513,6 +2513,18 @@ bool swift::swift_dynamicCast(OpaqueValue *dest,
     // unwrapping the target. This handles an optional source wrapped within an
     // existential that Optional conforms to (Any).
     if (auto srcExistentialType = dyn_cast<ExistentialTypeMetadata>(srcType)) {
+#if SWIFT_OBJC_INTEROP
+      // If coming from AnyObject, we may want to bridge.
```

Comment:
I suppose we could also support it from `NSNull`. `AnyObject` is the only Swift-level formal type that'd make sense to cast back to `Optional` otherwise.